test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs

```diff
@@ -140,5 +144,215 @@ public async Task RestoreAsync_WithMinimalProjectAndAdditionalErrorMessage_Write
+
+            // Assert
+            result.Should().HaveCount(1);
+            restoreSummary = result[0];
+            restoreSummary.Success.Should().BeTrue();
+            restoreSummary.NoOpRestore.Should().BeTrue();
+
+            progressReporter.StartProjectUpdateQueue.Count.Should().Be(0);
+            progressReporter.EndProjectUpdateQueue.Count.Should().Be(0);
+        }
+
+        [Fact]
+        public async Task RestoreAsync_WithProgressReporter_WithMultipleProjects_ProgressIsNotReportedForChangedProjetsOnly()
+        {
+            // Arrange
+            using var pathContext = new SimpleTestPathContext();
+            var settings = Settings.LoadDefaultSettings(pathContext.SolutionRoot);
+            var project1 = ProjectTestHelpers.GetPackageSpec(settings, projectName: "project1", rootPath: pathContext.SolutionRoot);
+            var project2 = ProjectTestHelpers.GetPackageSpec(settings, projectName: "project2", rootPath: pathContext.SolutionRoot);
+            var progressReporter = new RestoreProgressReporter();
+
+            // Pre-Conditions
+            IReadOnlyList<RestoreSummary> result = await DependencyGraphRestoreUtility.RestoreAsync(
+                ProjectTestHelpers.GetDGSpecFromPackageSpecs(project1, project2),
+                new DependencyGraphCacheContext(),
+                new RestoreCommandProvidersCache(),
+                cacheContextModifier: _ => { },
+                sources: new SourceRepository[0],
+                parentId: Guid.Empty,
+                forceRestore: false,
+                isRestoreOriginalAction: true,
+                additionalMessages: null,
+                progressReporter: progressReporter,
+                new TestLogger(),
+                CancellationToken.None);
+
+            // Pre-Conditions
+            result.Should().HaveCount(2);
+            foreach (RestoreSummary restoreSummary in result)
+            {
+                restoreSummary.Success.Should().BeTrue();
+                restoreSummary.NoOpRestore.Should().BeFalse();
+            }
+
+            progressReporter.StartProjectUpdateQueue.Count.Should().Be(2);
+            progressReporter.EndProjectUpdateQueue.Count.Should().Be(2);
+
+            var startUpdateProjects = new List<string>();
+
+            while (progressReporter.StartProjectUpdateQueue.Count > 0)
+            {
+                startUpdateProjects.Add(progressReporter.StartProjectUpdateQueue.Dequeue().Item1);
+            }
+
+            var endUpdateProjects = new List<string>();
+            while (progressReporter.EndProjectUpdateQueue.Count > 0)
+            {
+                endUpdateProjects.Add(progressReporter.EndProjectUpdateQueue.Dequeue().Item1);
+            }
+
+            startUpdateProjects.OrderedEquals(endUpdateProjects, e => e, PathUtility.GetStringComparerBasedOnOS(), PathUtility.GetStringComparerBasedOnOS()).Should().BeTrue("The start and end queues should have the same projects");
```

Comment:
Maybe I don't understand the test code well enough, but if projects are restored in parallel, doesn't that by definition make the order indeterministic?

Reply:
I'm checking OrderedEquals because of that.